lib/new_relic/telemetry/broadway.ex

```diff
@@ -0,0 +1,190 @@
+  end
+
+  def handle_event(@processor_start, _measurements, metadata, _config) do
+    Transaction.start_transaction("Broadway", processor_name(metadata))
+    NewRelic.add_attributes(processor_start_attributes(metadata))
+
+    DistributedTrace.generate_new_context()
+    |> DistributedTrace.track_transaction(transport_type: "HTTP")
+  end
+
+  def handle_event(@processor_stop, _measurements, metadata, _config) do
+    NewRelic.add_attributes(processor_stop_attributes(metadata))
+    Transaction.stop_transaction()
+  end
+
+  def handle_event(@message_stop, measurements, metadata, _config) do
+    track_message_segment(measurements, metadata)
+  end
+
+  def handle_event(@message_fail, measurements, metadata, _config) do
+    track_message_segment(measurements, metadata, %{
+      kind: metadata.kind,
+      reason: metadata.reason,
+      stack: metadata.stacktrace
+    })
+  end
+
+  def handle_event(@consumer_start, _measurements, metadata, _config) do
+    Transaction.start_transaction("Broadway", consumer_name(metadata))
+    NewRelic.add_attributes(consumer_start_attributes(metadata))
+  end
+
+  def handle_event(@consumer_stop, _measurements, metadata, _config) do
+    NewRelic.add_attributes(consumer_stop_attributes(metadata))
+    Transaction.stop_transaction()
+  end
+
+  defp extract_callback_module_name(name) do
+    Module.split(name) |> :lists.reverse() |> Enum.drop(2) |> :lists.reverse() |> Enum.join(".")
+  end
+
+  defp processor_name(%{name: name}) do
+    %{"processor_key" => processor_key} =
+      Regex.named_captures(~r/Processor_(?<processor_key>\w+)_\d+$/, to_string(name))
+
+    "#{extract_callback_module_name(name)}/Processor/#{processor_key}"
+  end
+
+  defp processor_start_attributes(%{name: name, messages: messages}) do
+    [
+      "broadway.module": extract_callback_module_name(name),
+      "broadway.stage": :processor,
+      "broadway.processor_message_count": length(messages)
+    ]
+  end
+
+  defp processor_stop_attributes(metadata) do
+    info = Process.info(self(), [:memory, :reductions])
+
+    [
+      "broadway.processor_successful_to_ack_count": length(metadata.successful_messages_to_ack),
+      "broadway.processor_successful_to_forward_count":
+        length(metadata.successful_messages_to_forward),
+      "broadway.processor_failed_count": length(metadata.failed_messages),
+      memory_kb: _bytes_to_kilobytes = info[:memory] / 1024,
+      reductions: info[:reductions]
+    ]
+  end
+
+  def track_message_segment(
+        %{duration: duration},
+        %{name: name, processor_key: processor_key},
+        error \\ nil
+      ) do
+    end_time_ms = System.system_time(:millisecond)
+    duration_ms = System.convert_time_unit(duration, :native, :millisecond)
+    duration_s = duration_ms / 1000
+    start_time_ms = end_time_ms - duration_ms
+
+    pid = inspect(self())
+    id = {:broadway_message, make_ref()}
+    parent_id = Process.get(:nr_current_span) || :root
```

Comment:
I'm not 100% sure we want to model this with a Transaction for the processor & segments for messages, but that aside, I think I'll want to make a helper function for fetching the "parent" / "id" stuff instead of duplicating this code in a few places, since it's not super easy...

Reply:
Good point, I'm not sure either. On my first tests looked a good idea, but with some more real word testing I'm seeing really huge transaction traces that may not be ideal to follow.

Reply:
Pushed a change here to make a single Transaction around each `handle_message/3`, so this code is note relevant anymore.